pkg/frontend/querymiddleware/cardinality_query_cache.go

```diff
@@ -0,0 +1,186 @@
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+const (
+	cardinalityLabelNamesQueryCachePrefix  = "cn:"
+	cardinalityLabelValuesQueryCachePrefix = "cv:"
+)
+
+type cardinalityQueryRequest interface {
+	// String returns a full representation of the request, used to uniquely identify
+	// a request in the cache.
+	String() string
+
+	// RequestType returns the type of the cardinality query, used to distinguish
+	// different type of requests in the cache.
+	RequestType() cardinality.RequestType
+}
+
+// cardinalityQueryCache is a http.RoundTripped wrapping the downstream with an HTTP response cache.
+// This RoundTripper is used to add caching support to cardinality analysis API endpoints.
+type cardinalityQueryCache struct {
+	cache  cache.Cache
+	limits Limits
+	next   http.RoundTripper
+	logger log.Logger
+}
+
+func newCardinalityQueryCacheRoundTripper(cache cache.Cache, limits Limits, next http.RoundTripper, logger log.Logger) http.RoundTripper {
+	return &cardinalityQueryCache{
+		cache:  cache,
+		limits: limits,
+		next:   next,
+		logger: logger,
+	}
+}
+
+// TODO metrics?
+func (c *cardinalityQueryCache) RoundTrip(req *http.Request) (*http.Response, error) {
+	ctx := req.Context()
+
+	spanLog, ctx := spanlogger.NewWithLogger(ctx, c.logger, "cardinalityQueryCache.RoundTrip")
+	defer spanLog.Finish()
+
+	// Skip the cache if disabled for this request.
+	if decodeCacheDisabledOption(req) {
+		level.Debug(spanLog).Log("msg", "cache disabled for the request")
+		return c.next.RoundTrip(req)
+	}
+
+	tenantIDs, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return nil, apierror.New(apierror.TypeBadData, err.Error())
+	}
+
+	// Skip the cache if disabled for the tenant.
+	cacheTTL := validation.SmallestPositiveNonZeroDurationPerTenant(tenantIDs, c.limits.ResultsCacheTTLForCardinalityQuery)
+	if cacheTTL <= 0 {
+		level.Debug(spanLog).Log("msg", "cache disabled for the tenant")
+		return c.next.RoundTrip(req)
+	}
+
+	// Decode the request.
+	queryReq, err := parseCardinalityQueryRequest(req)
+	if err != nil {
+		// Logging as info because it's not an actionable error here.
+		// We defer it to the downstream.
+		level.Info(spanLog).Log("msg", "skipped cardinality query caching because failed to parse the request", "err", err)
+
+		// We skip the caching but let the downstream try to handle it anyway,
+		// since it's not our responsibility to decide how to response in this case.
+		return c.next.RoundTrip(req)
+	}
+
+	// Lookup the cache.
+	cacheKey, hashedCacheKey := generateCardinalityQueryRequestCacheKey(tenantIDs, queryReq)
+	res := c.fetchCachedResponse(ctx, cacheKey, hashedCacheKey)
+	if res != nil {
+		level.Debug(spanLog).Log("msg", "response fetched from the cache")
+		return res, nil
+	}
+
+	res, err = c.next.RoundTrip(req)
+	if err != nil {
+		return res, err
+	}
+
+	// Store the result in the cache.
+	if isCardinalityQueryResponseCachable(res) {
+		// The res.Body will be replaced after calling the following function.
```

Comment:
I'd move this comment to `EncodeCachedHTTPResponse`, which should be documented as preserving `res.Body`.

Reply:
Good point.